sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1796,6 +1796,9 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
       case s: Sort if !s.resolved || s.missingInput.nonEmpty =>
         resolveReferencesInSort(s)
 
+      case u: UnresolvedWithCTERelations =>
```

Comment:
Thinking about it more, now we need to special-case `UnresolvedWithCTERelations` twice: once in `ResolveReferences` to resolve session variables and once in `ResolveIdentifierClause` to resolve identifier and look up CTE relations.

How about we make `UnresolvedWithCTERelations` an unary code, and only special case it once in `ResolveRelations` that we should look up from CTE relations for `UnresolvedRelations` inside`UnresolvedWithCTERelations`? Sorry for the back and forth!

Reply:
No worry!
Hm, issue with that approach is that ResolveRelations is traversing the tree in bottom up manner so we will first do table lookup instead of CTE relations lookup since it will first encounter UnresolvedRelation instead of UnresolvedWithCTERelations?

Reply:
you are right, it's better to keep the bottom-up resolotion.